File: adt2unit/pas/generate_custom_vibrato.pas
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

procedure generate_custom_vibrato(value: Byte);

const
  vibtab_size: array[0..15] of Byte = (
    16,16,16,16,32,32,32,32,64,64,64,64,128,128,128,128);

var
  mul_r: Real;
  mul_b: Byte;
  idx,idx2: Byte;

function min0(value: Longint): Longint;
begin
  If (value >= 0) then min0 := value
  else min0 := 0;
end;

begin
  Case value of
    // set default speed table
    0: begin
         vibtrem_table_size := def_vibtrem_table_size;
         Move(def_vibtrem_table,vibtrem_table,SizeOf(vibtrem_table));
       end;

    // set custom speed table (fixed size = 32)
    1..239:
       begin
         vibtrem_table_size := def_vibtrem_table_size;
         mul_r := value/16;
         For idx2 := 0 to 7 do
          begin
            vibtrem_table[idx2*32] := 0;
            For idx := 1 to 16 do
              vibtrem_table[idx2*32+idx] := ROUND(idx*mul_r);
            For idx := 17 to 31 do
              vibtrem_table[idx2*32+idx] := ROUND((32-idx)*mul_r);
          end;
       end;

    // set custom speed table (speed factor = 1-4)
    240..255:
       begin
         vibtrem_speed_factor := SUCC((value-240) MOD 4);
         vibtrem_table_size := 2*vibtab_size[value-240];
         mul_b := 256 DIV (vibtab_size[value-240]);
         For idx2 := 0 to PRED(128 DIV vibtab_size[value-240]) do
           begin
             vibtrem_table[2*vibtab_size[value-240]*idx2] := 0;
             For idx := 1 to vibtab_size[value-240] do
               vibtrem_table[2*vibtab_size[value-240]*idx2+idx] :=
                 min0(idx*mul_b-1);
             For idx := vibtab_size[value-240]+1 to
                        2*vibtab_size[value-240]-1 do
               vibtrem_table[2*vibtab_size[value-240]*idx2+idx] :=
                 min0((2*vibtab_size[value-240]-idx)*mul_b-1);
           end;
       end;
  end;
end;

File: adt2unit/generate_custom_vibrato.c
// This file is part of Adlib Tracker II (AT2).
//
// SPDX-FileType: SOURCE
// SPDX-FileCopyrightText: 2014-2024 The Adlib Tracker 2 Authors
// SPDX-License-Identifier: GPL-3.0-or-later

/*static*/ void generate_custom_vibrato (uint8_t value)
{
  if (value == 0)
  {
    // 0: set default speed table
    vibtrem_table_size = def_vibtrem_table_size;
    memmove (vibtrem_table, def_vibtrem_table, sizeof (vibtrem_table));
  }
  else if (value <= 239)
  {
    // 1..239: set custom speed table (fixed size = 32)
    float k = value / 16.0;
    unsigned i, n;
    uint8_t *a, *b;

    vibtrem_table_size = def_vibtrem_table_size;

    // Fill first 32 values

    vibtrem_table[0] = 0;

    for (i = 1; i <= 16; i++)
      vibtrem_table[i] = i * k;

    for (i = 17; i <= 31; i++)
      vibtrem_table[i] = (32 - i) * k;

    // Copy first 32 values to the rest

    a = vibtrem_table;
    b = &vibtrem_table[32];
    n = (sizeof (vibtrem_table) / sizeof (vibtrem_table[0])) - 32;
    do
    {
      *b = *a;
      a++;
      b++;
    } while (--n);
  }
  else
  {
    // 240..255: set custom speed table (speed factor = 1..4)
    unsigned l, k, i, n;

    value -= 240; // 0..15
    vibtrem_speed_factor = (value % 4) + 1; // 1..4
    l = 16 << (value >> 2); // 16, 32, 64, 128
    vibtrem_table_size = 2 * l; // 32, 64, 128, 256 // FIXME: 256 is truncated to 0
    k = (sizeof (vibtrem_table) / sizeof (vibtrem_table[0])) / l; // 2, 4, 8, 16

    // Fill first (2 * l) values

    vibtrem_table[0] = 0;

    for (i = 1; i <= l; i++)
    {
      int32_t x = i * k - 1;

      vibtrem_table[i] = (x >= 0) ? x : 0;
    }

    n = 2 * l - 1;
    for (i = l + 1; i <= n; i++)
    {
      int32_t x = (2 * l - i) * k - 1;

      vibtrem_table[i] = (x >= 0) ? x : 0;
    }

    // Copy first (2 * l) values to the rest

    n = (sizeof (vibtrem_table) / sizeof (vibtrem_table[0])) - 2 * l;
    if (n)
    {
      uint8_t *a = vibtrem_table, *b = &vibtrem_table[2 * l];

      do
      {
        *b = *a;
        a++;
        b++;
      } while (--n);
    }
  }
}
